### PR TITLE
Modify middleware listener to use stratigility pipe

### PR DESCRIPTION
**Pull request by asgrim**
Taking issue #196 into play, this PR proposes to update the `MiddlewareListener` to use `Stratigility` to allow piped middleware. This means the `middleware` param can become an `array` and everything is pushed into the `MiddlewarePipe`... as let's be fair, only allowing just a single middleware is only marginally useful.

For this to work, `zendframework/zend-stratigility` needs to become a requirement for those using `MiddlewareListener` (I added it as `require-dev` for now), so could be seen as a BC break potentially; so happy to negotiate on how we'd like this added to `composer.json`. Otherwise, I believe (pending CI results) all the other tests still pass though.

### CODE REVIEW

**Comment by asgrim** [comment]
Apparently I didn't have an up to date `develop` branch >.<

**Comment by asgrim** [comment]
Rebased correctly now \o/

File: src/MiddlewareListener.php [hunk 1]
```diff
@@ -64,7 +80,15 @@ public function onDispatch(MvcEvent $event)
             foreach ($routeMatch->getParams() as $key => $value) {
                 $psr7Request = $psr7Request->withAttribute($key, $value);
             }
-            $return = $middleware($psr7Request, Psr7Response::fromZend($response));
+            $return = $pipe(
+                $psr7Request,
+                $psr7ResponsePrototype,
+                function (PsrServerRequestInterface $request, PsrResponseInterface $response) {
+                    throw new ReachedFinalHandlerException(
```
Comment (Ocramius) [inline src/MiddlewareListener.php]: @weierophinney are you OK with moving this to a static named constructor instead? Asking because consistency...

File: src/MiddlewareListener.php [hunk 2]
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
...
-            return $return;
+        $psr7ResponsePrototype = Psr7Response::fromZend($response);
+        $pipe = new MiddlewarePipe();
+        $pipe->setResponsePrototype($psr7ResponsePrototype);
+
+        $middlewaresToBePiped = !is_array($middleware) ? [$middleware] : $middleware;
+
+        $middlewareName = 'noMiddlewarePiped';
+        $middlewareToBePiped = null;
+        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
```
Comment (Ocramius) [inline src/MiddlewareListener.php]: Segregate to a private method.

File: src/MiddlewareListener.php [hunk 3]
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
...
-        }
-        if (! is_callable($middleware)) {
-            $return = $this->marshalMiddlewareNotCallable($application::ERROR_MIDDLEWARE_CANNOT_DISPATCH, $middlewareName, $event, $application);
-            $event->setResult($return);
-            return $return;
+        $psr7ResponsePrototype = Psr7Response::fromZend($response);
+        $pipe = new MiddlewarePipe();
+        $pipe->setResponsePrototype($psr7ResponsePrototype);
+
+        $middlewaresToBePiped = !is_array($middleware) ? [$middleware] : $middleware;
```
Comment (Ocramius) [inline src/MiddlewareListener.php]: Don't negate: invert the ternary please

File: src/MiddlewareListener.php [hunk 4]
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
...
+
+        $middlewareName = 'noMiddlewarePiped';
+        $middlewareToBePiped = null;
+        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+            $middlewareName = is_string($middlewareToBePiped) ? $middlewareToBePiped : get_class($middlewareToBePiped);
```
Comment (Ocramius) [inline src/MiddlewareListener.php]: Because [`get_class()` is stupid](https://3v4l.org/5O4QA) this check is insufficient

File: src/MiddlewareListener.php [hunk 5]
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
...
+
+            if (is_string($middlewareToBePiped) && $serviceManager->has($middlewareToBePiped)) {
```
Comment (Ocramius) [inline src/MiddlewareListener.php]: `is_string()` check is duplicate

File: src/MiddlewareListener.php [hunk 6]
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
...
+                $middlewareToBePiped = $serviceManager->get($middlewareToBePiped);
+            }
+            if (! is_callable($middlewareToBePiped)) {
```
Comment (Ocramius) [inline src/MiddlewareListener.php]: Should probably also **NOT** be a string?

Comment (Ocramius) [reply src/MiddlewareListener.php]: An example is an incompatible function name being passed in.

File: src/MiddlewareListener.php (hunk 2 above)
Comment (Ocramius) [reply src/MiddlewareListener.php]: Probably should just return `$pipe` in the private method.

File: src/MiddlewareListener.php [hunk 7]
```diff
@@ -75,7 +99,9 @@ public function onDispatch(MvcEvent $event)
             $event->setName(MvcEvent::EVENT_DISPATCH_ERROR);
             $event->setError($application::ERROR_EXCEPTION);
             $event->setController($middlewareName);
-            $event->setControllerClass(get_class($middleware));
+            if (null !== $middlewareToBePiped) {
+                $event->setControllerClass(get_class($middlewareToBePiped));
```
Comment (Ocramius) [inline src/MiddlewareListener.php]: You always assume that the last piped middleware is the "responding" one here... Probably needs documenting/commenting

File: src/MiddlewareListener.php [hunk 8]
```diff
@@ -47,15 +50,28 @@ public function onDispatch(MvcEvent $event)
...
-            $event->setResult($return);
-            return $return;
+        $psr7ResponsePrototype = Psr7Response::fromZend($response);
+        $pipe = new MiddlewarePipe();
+        $pipe->setResponsePrototype($psr7ResponsePrototype);
+
+        $middlewaresToBePiped = !is_array($middleware) ? [$middleware] : $middleware;
+
+        $middlewareName = 'noMiddlewarePiped';
+        $middlewareToBePiped = null;
```
Comment (Ocramius) [inline src/MiddlewareListener.php]: Can an empty pipeline be considered valid?

**Review by Ocramius — DISMISSED** [review]

File: src/MiddlewareListener.php (hunk 8 above)
Comment (asgrim) [reply src/MiddlewareListener.php]: I don't see why not; you'll just reach the final handler anyway...

File: src/MiddlewareListener.php (hunk 4 above)
Comment (asgrim) [reply src/MiddlewareListener.php]: Didn't change this from the original code...

File: src/MiddlewareListener.php (hunk 6 above)
Comment (asgrim) [reply src/MiddlewareListener.php]: If it's a string, it won't be a string any more. Also just following same logic that was already there...

File: src/MiddlewareListener.php (hunk 7 above)
Comment (asgrim) [reply src/MiddlewareListener.php]: Removed these as there's no longer just one middleware, so doesn't make sense. I guess this causes a definite BC break which is kinda rubbish though?

**Comment by asgrim** [comment]
@Ocramius some changes applied as requested.

File: src/MiddlewareListener.php (hunk 5 above)
Comment (asgrim) [reply src/MiddlewareListener.php]: Not sure what to do about it; it's just the flexibility around using a string as the middleware (load from container), or a callable (not ideal, but allowed)

File: src/MiddlewareListener.php (hunk 2 above)
Comment (asgrim) [reply src/MiddlewareListener.php]: Done

File: src/MiddlewareListener.php (hunk 3 above)
Comment (asgrim) [reply src/MiddlewareListener.php]: Done

**Comment by asgrim** [comment]
*shakes fist at PHP 5.6 support*

File: composer.json [hunk 9]
```diff
@@ -27,7 +27,8 @@
         "zendframework/zend-json": "^2.6.1 || ^3.0",
         "zendframework/zend-psr7bridge": "^0.2",
         "fabpot/php-cs-fixer": "1.7.*",
-        "phpunit/phpunit": "^4.5"
+        "phpunit/phpunit": "^4.5",
+        "zendframework/zend-stratigility": "^2.0"
```
Comment (weierophinney) [inline composer.json]: This should likely be `^2.0.1`, but I can take care of that during merge.

File: src/MiddlewareListener.php (hunk 1 above)
Comment (weierophinney) [reply src/MiddlewareListener.php]: I am, but I'm not convinced it's needed, as there's a static message with no dynamic substitutions, and no additional behavior.

File: src/MiddlewareListener.php [hunk 10]
```diff
@@ -74,8 +97,6 @@ public function onDispatch(MvcEvent $event)
         if ($caughtException !== null) {
             $event->setName(MvcEvent::EVENT_DISPATCH_ERROR);
             $event->setError($application::ERROR_EXCEPTION);
-            $event->setController($middlewareName);
-            $event->setControllerClass(get_class($middleware));
```
Comment (weierophinney) [inline src/MiddlewareListener.php:83]: Why are these lines removed?

File: src/MiddlewareListener.php [hunk 11]
```diff
@@ -96,6 +117,42 @@ public function onDispatch(MvcEvent $event)
...
+            if (null === $middlewareToBePiped) {
+                throw new \InvalidArgumentException('Middleware name cannot be null');
+            }
+
+            $middlewareName = is_string($middlewareToBePiped) ? $middlewareToBePiped : get_class($middlewareToBePiped);
+
+            if (is_string($middlewareToBePiped) && $serviceLocator->has($middlewareToBePiped)) {
+                $middlewareToBePiped = $serviceLocator->get($middlewareToBePiped);
+            }
+            if (! is_callable($middlewareToBePiped)) {
```
Comment (weierophinney) [inline src/MiddlewareListener.php]: This needs to change to be:

```php
if (! $middlewareToBePiped instanceof \Http\Interop\ServerMiddleware\MiddlewareInterface
    && ! is_callable($middlewareToBePiped)
)
```

as zend-stratigility 2.0 allows either of those types. Testing only for `is_callable()` means that http-interop middleware cannot be used!

File: src/MiddlewareListener.php (hunk 1 above)
Comment (weierophinney) [reply src/MiddlewareListener.php]: One potential refactor I see here is to instead define `ReachedFinalHandlerException` as follows:

```php
class ReachedFinalHandlerException extends RuntimeException
{
    public static function createFinalHandler()
    {
        return new self('Reached the final handler for middleware pipe - check the pipe configuration');
    }

    public function __invoke()
    {
        throw $this;
    }
}
```

This would then allow you to do the following above:

```php
$return = $pipe($psr7Request, $psr7Response, ReachedFinalHandlerException::createFinalHandler());
```

Not sure if that's abusing inheritance and SRP, though. :grin:

File: src/MiddlewareListener.php (hunk 11 above)
Comment (weierophinney) [reply src/MiddlewareListener.php]: Once you make those changes, you'll also need to add tests for piping http-interop middleware.

**Review by weierophinney — CHANGES_REQUESTED** [review]

**Comment by weierophinney** [comment]
Also, @asgrim — You'll need to rebase against latest develop branch, as I pushed a bunch of QA tooling changes earlier (PHPUnit 5.7/6.0 support, in particular, as well as zend-coding-standards integration).

File: src/MiddlewareListener.php (hunk 10 above)
Comment (asgrim) [reply src/MiddlewareListener.php:83]: There is potentially no longer just "one" middleware name... how do we decide which is the failed one? First? Last? IIRC I don't think there's any way (outside of the exception bubble here)

Comment (weierophinney) [reply src/MiddlewareListener.php:83]: Ah, right - that makes sense, then.

File: src/MiddlewareListener.php (hunk 1 above)
Comment (asgrim) [reply src/MiddlewareListener.php]: Yikes, I'm not really sure about the exception itself being the final handler, although I gotta say it's quite clever... does feel a bit wrong. I'll just refactor to named constructor for now and if you want to change it to be the actual final handler, I'm happy to :)

File: test/MiddlewareListenerTest.php [hunk 12]
```diff
@@ -88,6 +92,38 @@ public function testSuccessfullyDispatchesMiddleware()
         $this->assertEquals('Test!', $return->getBody());
     }
 
+    public function testSuccessfullyDispatchesHttpInteropMiddleware()
+    {
+        $expectedOutput = uniqid('expectedOutput', true);
+
+        $event = $this->createMvcEvent('path', new class($expectedOutput) implements MiddlewareInterface {
```
Comment (asgrim) [inline test/MiddlewareListenerTest.php]: bah, doesn't work on PHP 5.6 - I thought `develop` branches were min PHP 7.1 now? Maybe I didn't read properly?

**Comment by asgrim** [comment]
@weierophinney all updated, thanks for review!

File: src/MiddlewareListener.php [hunk 13]
```diff
@@ -69,7 +80,13 @@ public function onDispatch(MvcEvent $event)
             foreach ($routeMatch->getParams() as $key => $value) {
                 $psr7Request = $psr7Request->withAttribute($key, $value);
             }
-            $return = $middleware($psr7Request, Psr7Response::fromZend($response));
+            $return = $pipe(
```
Comment (weierophinney) [inline src/MiddlewareListener.php]: @asgrim So sorry I didn't notice this earlier, but... If we're pinning to Stratigility 2, we should likely use the http-interop `process()` method here instead of `__invoke()`, which means using a `DelegateInterface` instead of a callable.

May I suggest the following:

```php
$pipe->setResponsePrototype($psr7ResponsePrototype);
$return = $pipe->process($psr7Request, new \Zend\Stratigility\Delegate\CallableDelegateDecorator(
    function (PsrServerRequestInterface $request, PsrResponseInterface $response) {
        throw ReachedFinalHandlerException::create();
    }
));
```

That would accomplish the same as what you have here, and be forwards compatible.

File: src/Exception/MiddlewareNotCallableException.php [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2017 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace Zend\Mvc\Exception;
+
+final class MiddlewareNotCallableException extends RuntimeException
```
Comment (weierophinney) [inline src/Exception/MiddlewareNotCallableException.php]: Since we're allowing either callable or http-interop middleware, perhaps we should change this to `InvalidMiddlewareException` instead?

File: src/MiddlewareListener.php [hunk 15]
```diff
@@ -101,6 +116,42 @@ public function onDispatch(MvcEvent $event)
...
+    private function createPipeFromSpec(
+        ContainerInterface $serviceLocator,
+        ResponseInterface $responsePrototype,
+        array $middlewaresToBePiped
+    ) {
+        $pipe = new MiddlewarePipe();
+        $pipe->setResponsePrototype($responsePrototype);
+        foreach ($middlewaresToBePiped as $middlewareToBePiped) {
+            if (null === $middlewareToBePiped) {
+                throw new \InvalidArgumentException('Middleware name cannot be null');
```
Comment (weierophinney) [inline src/MiddlewareListener.php]: If we rename the exception to `InvalidMiddlewareException`, we could also add another named constructor: `::becauseNull()` (obviously, choose something better!).

**Review by weierophinney — CHANGES_REQUESTED** [review]
Few last nit-picks, but looks great!

File: src/MiddlewareListener.php (hunk 13 above)
Comment (asgrim) [reply src/MiddlewareListener.php]: Yes, this seems reasonable, though I note that `__invoke` does the wrapping in the `CallableDelegateDecorator` for us, but I have no strong feelings either way. If `__invoke` is eventually going to go, then yes the switch makes sense!

`setResponsePrototype` is already called in `createPipeFromSpec` btw, so I don't think needs calling again (unless I missed something)

Comment (weierophinney) [reply src/MiddlewareListener.php]: Ah, okay, nice.

As noted, using `process()` will be forwards compatible, meaning that when Stratigility adopts a finalized PSR-15, we likely can just add an `||` to the constraint and not need any actual code changes. I figure, do it once. :smile:

**Review by weierophinney — APPROVED** [review]
Per discussion in Slack, I'll perform the last requested changes now and merge.

Thanks, @asgrim!

**Comment by weierophinney** [comment]
Thanks, @asgrim
